src/adapter/src/util.rs

```diff
@@ -285,3 +297,77 @@ impl ShouldHalt for StashError {
         self.is_fence()
     }
 }
+
+impl ShouldHalt for StorageError {
+    fn should_halt(&self) -> bool {
+        match self {
+            StorageError::UpdateBeyondUpper(_)
+            | StorageError::ReadBeforeSince(_)
+            | StorageError::InvalidUppers(_) => true,
+            StorageError::SourceIdReused(_)
+            | StorageError::IdentifierMissing(_)
+            | StorageError::ClientError(_)
+            | StorageError::DataflowError(_) => false,
+            StorageError::IOError(e) => e.should_halt(),
+        }
+    }
+}
+
+impl ShouldHalt for DataflowCreationError {
+    fn should_halt(&self) -> bool {
+        match self {
+            DataflowCreationError::SinceViolation(_) => true,
+            DataflowCreationError::InstanceMissing(_)
+            | DataflowCreationError::CollectionMissing(_)
+            | DataflowCreationError::MissingAsOf => false,
+        }
+    }
+}
+
+impl ShouldHalt for CollectionUpdateError {
+    fn should_halt(&self) -> bool {
+        match self {
+            CollectionUpdateError::InstanceMissing(_)
+            | CollectionUpdateError::CollectionMissing(_) => false,
+        }
+    }
+}
+
+impl ShouldHalt for PeekError {
+    fn should_halt(&self) -> bool {
+        match self {
+            PeekError::SinceViolation(_) => true,
+            PeekError::InstanceMissing(_)
+            | PeekError::CollectionMissing(_)
+            | PeekError::ReplicaMissing(_) => false,
+        }
+    }
+}
+
+impl ShouldHalt for SubscribeTargetError {
+    fn should_halt(&self) -> bool {
+        match self {
+            SubscribeTargetError::InstanceMissing(_)
+            | SubscribeTargetError::SubscribeMissing(_)
+            | SubscribeTargetError::ReplicaMissing(_)
+            | SubscribeTargetError::SubscribeAlreadyStarted => false,
+        }
+    }
+}
+
+impl ShouldHalt for TransformError {
+    fn should_halt(&self) -> bool {
+        match self {
+            TransformError::Internal(_)
+            | TransformError::LetRecUnsupported
+            | TransformError::IdentifierMissing(_) => false,
+        }
+    }
+}
+
+// Needed for some functions that return an () error. We have no info, so just panic.
+impl ShouldHalt for () {
+    fn should_halt(&self) -> bool {
+        false
+    }
+}
```

Comment:
@benesch No need to review this entire PR, but I was hoping that you could spot check these. As a rule, I called `unwrap_or_terminate` for all controller methods. The logic being, that even if some of them return no halt-able errors now, they may return halt-able errors in the future. That's why we have some implementations always returning `false`.